relayer_base/src/logging.rs

```diff
@@ -0,0 +1,220 @@
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::{fmt, Layer};
+use xrpl_api::Transaction;
+use xrpl_types::AccountId;
+
+pub fn setup_logging(config: &Config) -> (ClientInitGuard, SdkTracerProvider) {
+    let environment = std::env::var("ENVIRONMENT").unwrap_or_else(|_| "development".to_string());
+
+    let guard = sentry::init((
+        config.sentry_dsn.to_string(),
+        sentry::ClientOptions {
+            release: sentry::release_name!(),
+            environment: Some(std::borrow::Cow::Owned(environment)),
+            traces_sample_rate: 1.0,
+            ..Default::default()
+        },
+    ));
+
+    global::set_text_map_propagator(TraceContextPropagator::new());
+
+    let path = match env::current_exe() {
+        Ok(exe_path) => exe_path
+            .file_name()
+            .expect("missing file name")
+            .to_str()
+            .expect("missing file name")
+            .to_string(),
+        Err(_) => "<unknown>".to_string(),
+    };
+
+    let resource = Resource::builder().with_service_name(path).build();
+
+    let tracer_provider = if let Some(jaeger_url) = &config.jaeger_grpc_url {
+        let exporter = SpanExporter::builder()
+            .with_tonic()
+            .with_endpoint(jaeger_url)
+            .build()
+            .expect("Failed to create Jaeger exporter");
+
+        SdkTracerProvider::builder()
+            .with_sampler(Sampler::ParentBased(Box::new(Sampler::TraceIdRatioBased(
+                1.0,
+            ))))
+            .with_resource(resource)
+            .with_batch_exporter(exporter)
+            .build()
+    } else {
+        SdkTracerProvider::builder()
+            .with_sampler(Sampler::ParentBased(Box::new(Sampler::TraceIdRatioBased(
+                1.0,
+            ))))
+            .with_resource(resource)
+            .build()
+    };
+
+    let tracer = tracer_provider.tracer("relayer");
+
+    let fmt_layer = fmt::layer()
+        .with_target(true)
+        .with_filter(LevelFilter::DEBUG);
+
+    let sentry_layer = sentry::integrations::tracing::layer()
+        .event_filter(|metadata| match *metadata.level() {
+            Level::ERROR => EventFilter::Event,
+            Level::WARN => EventFilter::Event,
+            _ => EventFilter::Breadcrumb,
+        })
+        .span_filter(|_| false); // Disable all spans for Sentry
+
+    if config.jaeger_grpc_url.is_some() {
+        let otel_layer = tracing_opentelemetry::layer()
+            .with_tracer(tracer)
+            .with_filter(LevelFilter::INFO);
+
+        tracing_subscriber::registry()
+            .with(fmt_layer) // Console logging
+            .with(sentry_layer) // Sentry logging
+            .with(otel_layer) // Otel layer for Jaeger
+            .init();
+    } else {
+        tracing_subscriber::registry()
+            .with(fmt_layer) // Console logging
+            .with(sentry_layer) // Sentry logging
+            .init();
+    }
```

Comment:
Something along those lines:
```suggestion
    let mut tracing_subscriber_builder = tracing_subscriber::registry()
        .with(fmt_layer)
        .with(sentry_layer);
    if config.jaeger_grpc_url.is_some() {
        let otel_layer = tracing_opentelemetry::layer()
            .with_tracer(tracer)
            .with_filter(LevelFilter::INFO);

        tracing_subscriber::registry()
            .with(fmt_layer) // Console logging
            .with(sentry_layer) // Sentry logging
            .with(otel_layer) // Otel layer for Jaeger
            .init();
             tracing_subscriber_builder = tracing_subscriber_builder.with(otel_layer);
    }
    tracing_subscriber_builder.init();
```

Reply:
@themicp you probably meant something like: 

```rust
let registry = tracing_subscriber::registry()
    .with(fmt_layer)
    .with(sentry_layer);

if config.jaeger_grpc_url.is_some() {
    let otel_layer = tracing_opentelemetry::layer()
        .with_tracer(tracer)
        .with_filter(LevelFilter::INFO);

    registry = registry.with(otel_layer);
}
registry.init();
```

The problem is that .with() returns a new type. Consider this code: 

```rust
    let mut registry_without_otel = tracing_subscriber::registry()
        .with(fmt_layer)
        .with(sentry_layer);

    let otel_layer = tracing_opentelemetry::layer()
         .with_tracer(tracer)
         .with_filter(LevelFilter::INFO);

    let registry_with_otel = registry_without_otel.with(otel_layer);
```

`registry_without_otel` will be of type (something like): `Layered<SentryLayer<Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>, Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>`

while `registry_with_otel` will be of type: ` Layered<Filtered<OpenTelemetryLayer<Layered<SentryLayer<Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>, Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>, SdkTracer>, LevelFilter, Layered<SentryLayer<Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>, Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>>, Layered<SentryLayer<Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>, Layered<Filtered<Layer<Registry>, LevelFilter, Registry>, Registry>>>`

So, we can't keep the result in the same variable.

I had two possible solutions in mind

1. Try to potentially deal with pointers and dereferencing - but I tried that and it became unwieldly quickly
2. Do something like this: 

```rust
let registry = tracing_subscriber::registry()
    .with(fmt_layer)
    .with(sentry_layer);

if config.jaeger_grpc_url.is_some() {
    let otel_layer = tracing_opentelemetry::layer()
        .with_tracer(tracer)
        .with_filter(LevelFilter::INFO);

    registry.with(otel_layer).init();
} else {
    registry.init();
}
```

This looks cleaner, but it's less explicit about what we are doing, and if we add another optional layer at some point, then we have to check for every permutation. 

Please let me know if you see a more elegant solution that I'm missing.